python/plugins/processing/tools/postgis.py

```diff
@@ -138,41 +172,74 @@ def _quote(self, ident):
 
 class GeoDB:
 
+    @classmethod
+    def from_name(cls, conn_name):
+        uri = uri_from_name(conn_name)
+        return cls(uri=uri)
+
     def __init__(self, host=None, port=None, dbname=None, user=None,
-                 passwd=None):
+                 passwd=None, service=None, uri=None):
         # Regular expression for identifiers without need to quote them
         self.re_ident_ok = re.compile(r"^\w+$")
 
-        self.host = host
-        self.port = port
-        self.dbname = dbname
-        self.user = user
-        self.passwd = passwd
-
-        if self.dbname == '' or self.dbname is None:
-            self.dbname = self.user
-
-        try:
-            self.con = psycopg2.connect(self.con_info())
-        except psycopg2.OperationalError as e:
-            raise DbError(unicode(e))
+        if uri:
+            self.uri = uri
+        else:
+            self.uri = QgsDataSourceURI()
+            if service:
+                self.uri.setConnection(service, dbname, user, passwd)
+            else:
+                self.uri.setConnection(host, port, dbname, user, passwd)
+
+        conninfo = self.uri.connectionInfo(False)
+        err = None
+        for i in range(4):
+            expandedConnInfo = uri.connectionInfo(True)
+            try:
+                self.con = psycopg2.connect(expandedConnInfo.encode('utf-8'))
+                if err is not None:
```

Comment:
Yes, it is correct.

If the connection works at the first pass, err is None, credentials comes froms the original uri, or from QgsAuthManager, and we don't have to put credentials into QgsCrendential.

This line is there to avoid code redundancy we have in db_manager.